File: 프로그래머스/2/87946. 피로도/피로도.js




function solution(k, dungeons) {
    let res = 0;
    const visited = new Array(dungeons.length).fill(0)
    
    function dfs(hp, depth){
        for(let i=0; i<dungeons.length; i++){
            if(!visited[i] && dungeons[i][0]<=hp){
                visited[i] = 1;
                dfs(hp-dungeons[i][1], depth+1)
                visited[i] = 0;
            }
        }
        
        res = Math.max(res, depth)
    }
    
    dfs(k, 0)
    
    return res;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = 0
    N = len(dungeons)
    visited = [False] * N
    
    def dfs(k, cnt):
        nonlocal answer
        
        answer = max(answer, cnt)
        
        for i in range(N):
            need, cost = dungeons[i]
            if not visited[i] and k >= need:
                visited[i] = True
                dfs(k - cost, cnt + 1)
                visited[i] = False
                
    dfs(k, 0)
    
    
    return answer